pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/util/math"
+)
+
+const (
+	// Interval for updating resource (CPU/memory) utilization
+	resourceUtilizationUpdateInterval = time.Second
+
+	memPathV2 = "/sys/fs/cgroup/memory.current"
+	cpuPathV2 = "/sys/fs/cgroup/cpu.stat"
+	memPathV1 = "/sys/fs/cgroup/memory/memory.stat"
+	cpuPathV1 = "/sys/fs/cgroup/cpu/cpuacct.usage"
+)
+
+type utilizationScanner interface {
+	fmt.Stringer
+
+	// Scan returns CPU time in seconds and memory utilization in bytes, or an error.
+	Scan() (float64, uint64, error)
+}
+
+type cgroupV2Scanner struct {
+}
+
+func (s cgroupV2Scanner) String() string {
+	return "cgroup v2"
+}
+
+func (s cgroupV2Scanner) Scan() (float64, uint64, error) {
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	memR, err := readFileNoStat(memPathV2)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	var memUtil uint64
+	if _, err := fmt.Fscanf(memR, "%d", &memUtil); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", memPathV2)
+	}
+
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	cpuR, err := readFileNoStat(cpuPathV2)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	var name string
+	var cpuTime uint64
+	const usageUsec = "usage_usec"
+	for {
+		if _, err := fmt.Fscanf(cpuR, "%s %d\n", &name, &cpuTime); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPathV2)
+		}
+
+		if name == usageUsec {
+			// This is CPU time in microseconds
+			break
+		}
+	}
+	if name != usageUsec {
+		return 0, 0, fmt.Errorf("failed scanning %s", cpuPathV2)
+	}
+
+	return float64(cpuTime) / 1e06, memUtil, nil
+}
+
+func newCgroupV2Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	return cgroupV2Scanner{}, true
+}
+
+type cgroupV1Scanner struct {
+}
+
+func (s cgroupV1Scanner) String() string {
+	return "cgroup v1"
+}
+
+func (s cgroupV1Scanner) Scan() (float64, uint64, error) {
+	memR, err := readFileNoStat(memPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	// Summing RSS and cache for memory usage
```

Comment:
Why the cache? The proc just uses the RSS and I think it's more correct. We don't care about cache. We want to limit once the machine memory is under stress, the cache gets purged and our process memory utilization is then more accurately tracked by the RSS. See: https://github.com/grafana/mimir/issues/2466

Reply:
As we discussed elsewhere, I've removed the cgroup logic.